src/algebra/homology/short_exact/preadditive.lean

```diff
@@ -0,0 +1,315 @@
+{ epi :=
+  begin
+    obtain ⟨χ, hχ⟩ := h.right_split,
+    haveI : epi (χ ≫ g) := by { rw hχ, apply_instance },
+    exact epi_of_epi χ g,
+  end,
+  mono := h.mono,
+  exact := h.exact }
+
+end has_zero_morphisms
+
+section preadditive
+variables [preadditive 𝒜]
+
+/-- An exact sequence `A -f⟶ B -g⟶ C` is *split* if there exist
+`φ : B ⟶ A` and `χ : C ⟶ B` such that:
+* `f ≫ φ = 𝟙 A`
+* `χ ≫ g = 𝟙 C`
+* `f ≫ g = 0`
+* `χ ≫ φ = 0`
+* `φ ≫ f + g ≫ χ = 𝟙 B`
+
+Such a sequence is automatically short exact (i.e., `f` is mono and `g` is epi). -/
+structure split : Prop :=
+(split : ∃ (φ : B ⟶ A) (χ : C ⟶ B),
+  f ≫ φ = 𝟙 A ∧ χ ≫ g = 𝟙 C ∧ f ≫ g = 0 ∧ χ ≫ φ = 0 ∧ φ ≫ f + g ≫ χ = 𝟙 B)
+
+variables [has_kernels 𝒜] [has_images 𝒜]
+
+lemma exact_of_split {A B C : 𝒜} (f : A ⟶ B) (g : B ⟶ C) (χ : C ⟶ B) (φ : B ⟶ A)
+  (hfg : f ≫ g = 0) (H : φ ≫ f + g ≫ χ = 𝟙 B) : exact f g :=
+{ w := hfg,
+  epi :=
+  begin
+    let ψ : (kernel_subobject g : 𝒜) ⟶ image_subobject f :=
+      subobject.arrow _ ≫ φ ≫ factor_thru_image_subobject f,
+    suffices : ψ ≫ image_to_kernel f g hfg = 𝟙 _,
+    { convert epi_of_epi ψ _, rw this, apply_instance },
+    rw ← cancel_mono (subobject.arrow _), swap, { apply_instance },
+    simp only [image_to_kernel_arrow, image_subobject_arrow_comp, category.id_comp, category.assoc],
+    calc (kernel_subobject g).arrow ≫ φ ≫ f
+        = (kernel_subobject g).arrow ≫ 𝟙 B : _
+    ... = (kernel_subobject g).arrow        : category.comp_id _,
+    rw [← H, preadditive.comp_add],
+    simp only [add_zero, zero_comp, kernel_subobject_arrow_comp_assoc],
+  end }
+
+section
+
+variables {f g}
+
+lemma split.exact (h : split f g) : exact f g :=
+by { obtain ⟨φ, χ, -, -, h1, -, h2⟩ := h, exact exact_of_split f g χ φ h1 h2 }
+
+lemma split.left_split (h : split f g) : left_split f g :=
+{ left_split := by { obtain ⟨φ, χ, h1, -⟩ := h, exact ⟨φ, h1⟩, },
+  epi := begin
+    obtain ⟨φ, χ, -, h2, -⟩ := h,
+    have : epi (χ ≫ g), { rw h2, apply_instance },
+    exactI epi_of_epi χ g,
+  end,
+  exact := h.exact }
+
+lemma split.right_split (h : split f g) : right_split f g :=
+{ right_split := by { obtain ⟨φ, χ, -, h1, -⟩ := h, exact ⟨χ, h1⟩, },
+  mono := begin
+    obtain ⟨φ, χ, h1, -⟩ := h,
+    have : mono (f ≫ φ), { rw h1, apply_instance },
+    exactI mono_of_mono f φ,
+  end,
+  exact := h.exact }
+
+lemma split.short_exact (h : split f g) : short_exact f g :=
+h.left_split.short_exact
+
+end
+
+lemma split.map {𝒜 ℬ : Type*} [category 𝒜] [preadditive 𝒜] [category ℬ] [preadditive ℬ]
+  (F : 𝒜 ⥤ ℬ) [functor.additive F] {A B C : 𝒜} {f : A ⟶ B} {g : B ⟶ C} (h : split f g) :
+  split (F.map f) (F.map g) :=
+begin
+  obtain ⟨φ, χ, h1, h2, h3, h4, h5⟩ := h,
+  refine ⟨⟨F.map φ, F.map χ, _⟩⟩,
+  simp only [← F.map_comp, ← F.map_id, ← F.map_add, F.map_zero, *, eq_self_iff_true, and_true],
+end
+
+/-- The sequence `A ⟶ A ⊞ B ⟶ B` is exact. -/
+lemma exact_inl_snd [has_binary_biproducts 𝒜] (A B : 𝒜) :
```

Comment:
We probably also want that this is `short_exact` and split. (And similarly below.)

Reply:
This can happen in a later PR. Extracting material from `for_mathlib` is excruciating enough as is.

Reply:
Or, anyone can add it. :-)